// @glimpse/glimpse-common/release/src/errors/ErrorCodes.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Gets a slug for the given error code
 */
function getSlugForErrorCode(errorCode) {
    return ErrorCode[errorCode];
}
exports.getSlugForErrorCode = getSlugForErrorCode;
/**
 * Get's a fully qualified URL for the given slug. Currently, this returns the location
 * on the Home repo's README, but will be modified once the main website is live for the
 * new version of Glimpse
 *
 * @parameter {string} slug - The slug to get the URL for, e.g. "package--version-support"
 * @return {string} The fully qualified URL for the slug
 */
function getUrlForErrorCode(errorCode) {
    switch (errorCode) {
        case ErrorCode.NewerUnsupportedPackageRequired:
        case ErrorCode.OlderUnsupportedPackageRequired:
            return 'https://github.com/Glimpse/Home#package--version-support';
        default:
            return 'https://github.com/Glimpse/Home';
    }
}
exports.getUrlForErrorCode = getUrlForErrorCode;
var ErrorCode;
(function (ErrorCode) {
    // Agent Errors
    ErrorCode[ErrorCode["GlimpseAgentErrorBase"] = 1000] = "GlimpseAgentErrorBase";
    ErrorCode[ErrorCode["NoContext"] = 1001] = "NoContext";
    ErrorCode[ErrorCode["UnexpectedContextValue"] = 1002] = "UnexpectedContextValue";
    ErrorCode[ErrorCode["PackageRequiredBeforeInit"] = 1003] = "PackageRequiredBeforeInit";
    ErrorCode[ErrorCode["OlderUnsupportedPackageRequired"] = 1004] = "OlderUnsupportedPackageRequired";
    ErrorCode[ErrorCode["NewerUnsupportedPackageRequired"] = 1005] = "NewerUnsupportedPackageRequired";
    ErrorCode[ErrorCode["HttpClientError"] = 1006] = "HttpClientError";
    ErrorCode[ErrorCode["HttpServerError"] = 1007] = "HttpServerError";
    ErrorCode[ErrorCode["StackHelperUnsupportedSourceMapUri"] = 1008] = "StackHelperUnsupportedSourceMapUri";
    ErrorCode[ErrorCode["StackHelperUnsupportedStackFrameFormat"] = 1009] = "StackHelperUnsupportedStackFrameFormat";
    ErrorCode[ErrorCode["HttpServerEarlyRequestTerminationError"] = 1010] = "HttpServerEarlyRequestTerminationError";
    ErrorCode[ErrorCode["AsyncTrackError"] = 1011] = "AsyncTrackError";
    ErrorCode[ErrorCode["AsyncTrackWarning"] = 1012] = "AsyncTrackWarning";
    // Server Errors
    ErrorCode[ErrorCode["GlimpseServerErrorBase"] = 5000] = "GlimpseServerErrorBase";
    ErrorCode[ErrorCode["AuthorizationInvocationFailed"] = 5001] = "AuthorizationInvocationFailed";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
