// @glimpse/glimpse-server/release/src/resources/ContextResource.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var MessageUtilities_1 = require("../messaging/MessageUtilities");
var Resource = (function () {
    function Resource(server) {
        this.name = 'context';
        this.uriTemplate = '?contextId={contextId}{&types}';
        this.templateName = 'contextTemplate';
        this.type = 'client';
        if (server) {
            this.init(server.providers.messageQuery);
        }
    }
    Resource.prototype.init = function (messageQuery) {
        this.messageQuery = messageQuery;
    };
    Resource.prototype.invoke = function (req, res) {
        if (!req.query.contextId) {
            res.sendStatus(404);
            res.send('missing contextId parameter');
        }
        else {
            var types = [];
            if (req.query.types) {
                types = req.query.types.split(',');
            }
            var messages = this.messageQuery.queryMessages(req.query.contextId, types);
            var payloads = MessageUtilities_1.MessageUtilities.toPayloadJson(messages);
            res.status(200);
            res.type('application/json');
            res.send(payloads);
        }
    };
    ;
    return Resource;
}());
exports.Resource = Resource;
