// @glimpse/glimpse-server/release/src/resources/MessageDebugResource.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var MessageUtilities_1 = require("../messaging/MessageUtilities");
var Resource = (function () {
    function Resource(server) {
        this.name = 'message-debug';
        this.uriTemplate = '{?types,contextIds}';
        this.type = 'client';
        if (server) {
            this.init(server.providers.messageQuery);
        }
    }
    Resource.prototype.init = function (messageQuery) {
        this.messageQuery = messageQuery;
    };
    Resource.prototype.invoke = function (req, res) {
        var types = req.query.types ? req.query.types.split(',') : [];
        var contextIds = req.query.contextIds ? req.query.contextIds.split(',') : undefined;
        var messages = [];
        if (contextIds) {
            for (var i = 0; i < contextIds.length; i++) {
                messages = messages.concat(this.messageQuery.queryMessages(contextIds[i], types));
            }
        }
        else {
            messages = this.messageQuery.queryMessages(undefined, types);
        }
        var payloads = MessageUtilities_1.MessageUtilities.toPayloadJson(messages);
        res.status(200);
        res.type('application/json');
        res.send(payloads);
    };
    ;
    return Resource;
}());
exports.Resource = Resource;
