// @glimpse/glimpse-server/release/src/resources/MetadataResource.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var hash_1 = require("../services/hash");
var UriTemplate_1 = require("./UriTemplate");
var _ = require("lodash");
var Resource = (function () {
    function Resource(server) {
        this.name = 'metadata';
        this.uriTemplate = '?hash={hash}';
        this.templateName = 'metadataTemplate';
        this.type = 'client';
        this.configSettings = server.providers.configSettings;
        this.resourceManager = server.providers.resourceManager;
        this.versionInfoService = server.providers.versionInfoService;
    }
    Resource.prototype.invoke = function (req, res) {
        var baseUri = UriTemplate_1.UriTemplate.getBaseUri(this.configSettings, req);
        var resources = _(this.resourceManager.resources)
            .reduce(function (result, resource) {
            result[resource.name] = UriTemplate_1.UriTemplate.fromResource(baseUri, resource);
            return result;
        }, {});
        var resourceHash = hash_1.Hash.hashObject(resources);
        var metadata = {
            'versions': this.versionInfoService.allInfo,
            'resources': resources,
            'hash': resourceHash,
            'experimentalMode': this.configSettings.getBoolean('enable.experimental.features', false)
        };
        res.json(metadata);
    };
    ;
    return Resource;
}());
exports.Resource = Resource;
