// @glimpse/glimpse-server/release/src/resources/ResourceAuthorization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ResourceAuthorization = (function () {
    function ResourceAuthorization() {
        this._agentAuthorizations = [];
        this._clientAuthorizations = [];
    }
    ResourceAuthorization.prototype.addAgentAuthorization = function (authorization) {
        this._agentAuthorizations.push(authorization);
    };
    ResourceAuthorization.prototype.addClientAuthorization = function (authorization) {
        this._clientAuthorizations.push(authorization);
    };
    ResourceAuthorization.prototype.canExecute = function (req, resourceType, callback) {
        switch (resourceType) {
            case 'agent':
                this.canExecuteType(req, this._agentAuthorizations, callback);
                break;
            case 'client':
                this.canExecuteType(req, this._clientAuthorizations, callback);
                break;
            default:
                callback(new Error("Unable to execute a request for the unrecognized resource type '" + resourceType + "'."), /* canExecute: */ false);
                break;
        }
    };
    ResourceAuthorization.prototype.canExecuteType = function (req, authorizations, callback) {
        if (authorizations.length === 0) {
            callback(/* error: */ undefined, /* canExecute: */ true);
        }
        else {
            var authorizationsCopy_1 = authorizations.slice();
            var next = function next() {
                var authorization = authorizationsCopy_1.shift();
                if (authorization) {
                    authorization(req, function (err, canExecute) {
                        if (err || !canExecute) {
                            callback(err, canExecute);
                        }
                        else {
                            next();
                        }
                    });
                }
                else {
                    callback(/* err: */ undefined, /* canExecute: */ true);
                }
            };
            next();
        }
    };
    return ResourceAuthorization;
}());
exports.ResourceAuthorization = ResourceAuthorization;
