// @glimpse/glimpse-agent-node/release/messaging/ContextManagerAsyncTrack.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ContextManagerBase_1 = require("./ContextManagerBase");
var async_track_1 = require("./../async-track/async-track");
var glimpse_common_1 = require("@glimpse/glimpse-common");
//tslint:disable-next-line:no-any
//const rawDebug = (<any>process)._rawDebug;
var ContextManagerAsyncTrack = (function (_super) {
    __extends(ContextManagerAsyncTrack, _super);
    function ContextManagerAsyncTrack() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContextManagerAsyncTrack.prototype.init = function (asyncTrack) {
        var self = this;
        _super.prototype.init.call(this);
        if (!asyncTrack) {
            asyncTrack = async_track_1.getAsyncTrack();
        }
        this.asyncTrack = asyncTrack;
        var handler = {
            onAsyncTransition: function onAsyncTransition(parentId, id) {
                //rawDebug(`onAsyncTransition: parentId ${parentId}, id ${id}`);
                var context = self.currentContext();
                if (context) {
                    return self.createAsyncState(context);
                }
                else {
                    return undefined;
                }
            },
            onBeforeInvocation: function onBeforeInvocation(id) {
                //rawDebug(`onBeforeInvocation:  id ${id}`);
            },
            onAfterInvocation: function onAfterInvocation(id) {
                //rawDebug(`onAfterInvocation:  id ${id}`);
            },
            onError: function onError(msg) {
                if (this.getErrorReportingService()) {
                    this.getErrorReportingService().reportError(glimpse_common_1.createAsyncTrackError(msg));
                }
            },
            onWarning: function onWarning(msg) {
                if (this.errorReportingService) {
                    this.getErrorReportingService().reportError(glimpse_common_1.createAsyncTrackWarning(msg));
                }
            }
        };
        this.asyncTrack.addHandler(handler);
    };
    ;
    ContextManagerAsyncTrack.prototype.createAsyncState = function (context) {
        return {
            GLIMPSE_NAMESPACE: {
                GLIMPSE_CONTEXT: context
            }
        };
    };
    // tslint:disable-next-line:no-any
    ContextManagerAsyncTrack.prototype.runInContext = function (context, callback) {
        var callbackWrapper = function callbackWrapper() {
            return callback(context);
        };
        return this.asyncTrack.runInContext(callbackWrapper, this.createAsyncState(context));
    };
    // tslint:disable-next-line:no-any
    ContextManagerAsyncTrack.prototype.runInNewContext = function (req, callback) {
        return this.runInContext(this.createContext(req), callback);
    };
    // tslint:disable-next-line:no-any
    ContextManagerAsyncTrack.prototype.runInNullContext = function (callback) {
        return this.runInContext(undefined, callback);
    };
    ContextManagerAsyncTrack.prototype.wrapInCurrentContext = function (callback) {
        var self = this;
        var asyncContext = this.asyncTrack.getCurrentContext();
        var wrapper = function wrapper() {
            var outerArgs = arguments;
            var outerThis = this;
            var callbackWrapper = function callbackWrapper() {
                callback.apply(outerThis, outerArgs);
            };
            return self.asyncTrack.runInContext(callbackWrapper, asyncContext);
        };
        return wrapper;
    };
    ContextManagerAsyncTrack.prototype.currentContext = function () {
        var asyncContext = this.asyncTrack.getCurrentContext();
        var glimpseContext = undefined;
        if (asyncContext) {
            var glimpseNS = asyncContext[ContextManagerAsyncTrack.GLIMPSE_NAMESPACE];
            if (glimpseNS) {
                glimpseContext = glimpseNS[ContextManagerAsyncTrack.GLIMPSE_CONTEXT];
            }
        }
        return glimpseContext;
    };
    return ContextManagerAsyncTrack;
}(ContextManagerBase_1.ContextManagerBase));
ContextManagerAsyncTrack.GLIMPSE_NAMESPACE = 'GLIMPSE_NAMESPACE';
ContextManagerAsyncTrack.GLIMPSE_CONTEXT = 'GLIMPSE_CONTEXT';
exports.ContextManagerAsyncTrack = ContextManagerAsyncTrack;
