// @glimpse/glimpse-agent-node/release/inspectors/ExpressInspectorMiddleware.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var MiddlewareWrapper_1 = require("./util/MiddlewareWrapper");
var Tracing_1 = require("./../tracing/Tracing");
var ExpressEvents_1 = require("./../tracing/module_instrumentors/ExpressEvents");
var ExpressInspectorMiddleware = (function () {
    function ExpressInspectorMiddleware() {
    }
    ExpressInspectorMiddleware.prototype.init = function (broker, providers) {
        var _this = this;
        this.broker = broker;
        this.dateTime = providers.dateTime;
        this.stackHelper = providers.stackHelper;
        this.configSettings = providers.configSettings;
        this.contextManager = providers.contextManager;
        this.middlewareWrapper = new MiddlewareWrapper_1.MiddlewareWrapper(this.stackHelper, providers.messageConverter, this.broker, this.contextManager, this.dateTime);
        Tracing_1.default.onAlways(ExpressEvents_1.EVENT_EXPRESS_INVOKE_PRE_ROUTE_METHOD, function (event) {
            _this.onMiddlewareMethodInvoked(event.data.originalThis, event.data.originalArgs, event.data.methodName, event.data.proxyFunction);
        });
        Tracing_1.default.onAlways(ExpressEvents_1.EVENT_EXPRESS_INVOKE_PRE_ROUTER_USE, function (event) {
            _this.onUseInvokedBegin(event.data.originalThis, event.data.originalArgs, event.data.proxyFunction);
        });
        Tracing_1.default.onAlways(ExpressEvents_1.EVENT_EXPRESS_INVOKE_PRE_ROUTER, function (event) {
            _this.onRouterInvokedEnd(event.data.router);
        });
        Tracing_1.default.onAlways(ExpressEvents_1.EVENT_EXPRESS_INVOKE_PRE_STATIC, function (event) {
            _this.onStaticInvokedEnd(event.data.middleware);
        });
    };
    ExpressInspectorMiddleware.prototype.onMiddlewareMethodInvoked = function (originalThis, originalArgs, method, proxyFunction) {
        if (originalThis.glimpse && originalThis.glimpse.ignore) {
            return;
        }
        if (originalArgs && originalArgs.length >= 1) {
            var frame = this.stackHelper.tryGetFirstUserCodeFrame(this.stackHelper.captureStack(proxyFunction, 100));
            var registrationCallstack = frame ? [frame] : [];
            this.wrapNestedMiddleware(originalArgs, [originalThis.path], method !== 'all' ? method.toUpperCase() : undefined, registrationCallstack);
        }
    };
    ExpressInspectorMiddleware.prototype.onUseInvokedBegin = function (originalThis, originalArgs, proxyFunction) {
        if (originalThis.glimpse && originalThis.glimpse.ignore) {
            return;
        }
        if (originalArgs && originalArgs.length >= 1) {
            var paths = ['/'];
            if (originalArgs.length > 1) {
                var firstArg = originalArgs[0];
                if (Array.isArray(firstArg) && firstArg.length > 0 && typeof firstArg[0] === 'string') {
                    paths = firstArg;
                }
                else if (typeof firstArg === 'string') {
                    paths = [firstArg];
                }
                else if (firstArg instanceof RegExp) {
                    paths = [firstArg.toString()];
                }
            }
            else if (originalArgs.length === 1) {
                var middleware = originalArgs[0];
                if (typeof middleware === 'function') {
                    if (!originalThis.glimpse) {
                        originalThis.glimpse = {
                            firstUse: middleware
                        };
                    }
                    else if (!originalThis.glimpse.firstUse) {
                        originalThis.glimpse.firstUse = middleware;
                    }
                    else if (!originalThis.glimpse.secondUse) {
                        originalThis.glimpse.secondUse = middleware;
                        // NOTE: If the first and second middleware added to a Router are 'query' and 'expressInit',
                        //       respectively, it's relatively certain that it's the built-in Express middleware that
                        //       we can't otherwise patch in order to attach metadata, so we do so here.
                        if (originalThis.glimpse.firstUse.name === 'query' && originalThis.glimpse.secondUse.name === 'expressInit') {
                            MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(originalThis.glimpse.firstUse, 'query', 'Express Query Parser', 'express');
                            MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(originalThis.glimpse.secondUse, 'expressInit', 'Express Initialization', 'express');
                        }
                    }
                }
            }
            // capture location of where use() is being called.
            var frame = this.stackHelper.tryGetFirstUserCodeFrame(this.stackHelper.captureStack(proxyFunction, 100));
            var registrationCallstack = frame ? [frame] : [];
            this.wrapNestedMiddleware(originalArgs, paths, undefined, registrationCallstack);
        }
    };
    ExpressInspectorMiddleware.prototype.wrapNestedMiddleware = function (array, paths, method, registrationCallstack) {
        for (var i = 0; i < array.length; i++) {
            var arg = array[i];
            if (typeof arg === 'function') {
                if (arg.glimpse && arg.glimpse.ignore) {
                    continue;
                }
                array[i] = this.middlewareWrapper.wrap(paths, method, arg, registrationCallstack);
            }
            else if (Array.isArray(arg)) {
                this.wrapNestedMiddleware(arg, paths, method, registrationCallstack);
            }
        }
    };
    ExpressInspectorMiddleware.prototype.onRouterInvokedEnd = function (router) {
        MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(router, 'router', 'Express Router', 'express');
    };
    ExpressInspectorMiddleware.prototype.onStaticInvokedEnd = function (middleware) {
        MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(middleware, 'serveStatic', 'Express Static File Server', 'express');
    };
    return ExpressInspectorMiddleware;
}());
exports.ExpressInspectorMiddleware = ExpressInspectorMiddleware;
