// @glimpse/glimpse-agent-node/release/inspectors/KnexProxy.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ProxyBase_1 = require("./ProxyBase");
var KnexProxy = (function (_super) {
    __extends(KnexProxy, _super);
    function KnexProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    KnexProxy.prototype.setupProxy = function (agent, knex) {
        return function knexProxy() {
            var realKnex = knex.apply(this, arguments);
            var onStart = function onKnexStart(builder) {
                var context = agent.providers.contextManager.currentContext();
                // NOTE: If there is no current context, the queries are being made outside of any outstanding
                //       HTTP request (such as during application startup) and these should not be logged.
                if (context) {
                    var startTime_1 = new Date();
                    var queries_1 = [];
                    builder.on('query', function onKnexQuery(query) {
                        queries_1.push(query);
                        agent.broker.createAndSendMessage({
                            commandText: query.sql,
                            commandType: '',
                            commandParameters: '',
                            commandMethod: '',
                            commandIsAsync: true,
                            commandStartTime: startTime_1
                        }, ['before-execute-command'], undefined /*indices*/, context);
                    });
                    builder.on('end', function onKnexEnd() {
                        var endTime = new Date();
                        queries_1.forEach(function forQuery() {
                            agent.broker.createAndSendMessage({
                                commandHadException: false,
                                commandEndTime: endTime,
                                commandDuration: endTime.getTime() - startTime_1.getTime(),
                                commandOffset: startTime_1.getTime() - context.startTime.getUnixTimestamp()
                            }, ['after-execute-command'], undefined /*indices*/, context);
                        });
                    });
                }
            };
            realKnex.client.on('start', onStart);
            return realKnex;
        };
    };
    ;
    Object.defineProperty(KnexProxy.prototype, "moduleName", {
        get: function () { return 'knex'; },
        enumerable: true,
        configurable: true
    });
    ;
    KnexProxy.prototype.init = function (agent, knex) {
        return this.setupProxy(agent, knex);
    };
    ;
    Object.defineProperty(KnexProxy.prototype, "forceLoadModule", {
        get: function () {
            return false;
        },
        enumerable: true,
        configurable: true
    });
    return KnexProxy;
}(ProxyBase_1.ProxyBase));
exports.KnexProxy = KnexProxy;
