// @glimpse/glimpse-agent-node/release/util/HttpHelper.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable:no-any */
var _ = require("lodash");
var parse = require("parseurl");
var cookie = require("cookie");
var MIME_TYPES_TO_CAPTURE = [
    /^text\//,
    /^application\/.*?xml/,
    /^application\/json/,
    /^application\/javascript/,
    /^application\/x-www-form-urlencoded/,
    /^multipart\/form-data/
];
var RequestHelper = (function () {
    function RequestHelper() {
    }
    RequestHelper.header = function (req, name) {
        var lc = name.toLowerCase();
        switch (lc) {
            case 'referer':
            case 'referrer':
                return req.headers.referrer || req.headers.referer;
            default:
                return req.headers[lc];
        }
    };
    /*
    // TODO: not worring about trust/proxies, etc. at the moment
    protocol: function(req) {
        ...
        var trust = this.app.get('trust proxy fn');
        if (!trust(req.connection.remoteAddress, 0)) {
            return proto;
        }
        ....
    },
    */
    RequestHelper.protocol = function (req) {
        var proto = req.connection.encrypted ? 'https' : 'http';
        proto = this.header(req, 'X-Forwarded-Proto') || proto;
        return proto.split(/\s*,\s*/)[0];
    };
    RequestHelper.parseurl = function (req) {
        return parse(req);
    };
    /*
    // TODO: not worring about trust/procies etc atm
    hostname: function(req) {
        ...
        var trust = this.app.get('trust proxy fn');
        var host = request.header(req, 'X-Forwarded-Host');
        if (!host || !trust(req.connection.remoteAddress, 0)) {
            host = request.header(req, 'Host');
        }
        ...
    }
    */
    RequestHelper.host = function (req) {
        return this.header(req, 'X-Forwarded-Host') || this.header(req, 'Host');
    };
    RequestHelper.hostname = function (req) {
        var host = RequestHelper.host(req);
        if (!host) {
            return;
        }
        // IPv6 literal support
        var offset = host[0] === '[' ? host.indexOf(']') + 1 : 0;
        var index = host.indexOf(':', offset);
        return index !== -1 ? host.substring(0, index) : host;
    };
    RequestHelper.parseCookies = function (req) {
        var result = this.header(req, 'cookie');
        return result ? cookie.parse(result) : undefined;
    };
    return RequestHelper;
}());
exports.RequestHelper = RequestHelper;
var ResponseHelper = (function () {
    function ResponseHelper() {
    }
    ResponseHelper.appendHeader = function (res, field, val) {
        var prev = res.getHeader(field);
        var value = val;
        if (prev) {
            // concat the new and prev vals
            value = Array.isArray(prev) ? prev.concat(val)
                : Array.isArray(val) ? [prev].concat(val)
                    : [prev, val];
        }
        return res.setHeader(field, value);
    };
    ResponseHelper.setCookie = function (res, name, value, options) {
        var opts = _.defaults({}, options, { path: '/' });
        if ('maxAge' in opts) {
            opts.expires = new Date(Date.now() + opts.maxAge);
            // TODO: What is this doing?
            opts.maxAge /= 1000;
        }
        this.appendHeader(res, 'Set-Cookie', cookie.serialize(name, String(value), opts));
        return this;
    };
    ResponseHelper.clearCookie = function (res, name, options) {
        var opts = _.defaults({}, options, { expires: new Date(1), path: '/' });
        return this.setCookie(res, name, '', opts);
    };
    return ResponseHelper;
}());
exports.ResponseHelper = ResponseHelper;
var HttpHelper = (function () {
    function HttpHelper() {
    }
    /**
     * helper method to store the context on an HTTP request or response instance
     */
    HttpHelper.setContext = function (requestResponse, context) {
        /*tslint:disable:no-any */
        requestResponse.__glimpseContext = context;
        /*tslint:enable:no-any */
    };
    /**
     * helper method to retrieve the context from an HTTP request or response instance
     */
    HttpHelper.getContext = function (requestResponse) {
        if (requestResponse) {
            /*tslint:disable:no-any */
            return (requestResponse.__glimpseContext);
            /*tslint:enable:no-any */
        }
        return undefined;
    };
    HttpHelper.isInProcessServerRequest = function (request) {
        // tslint:disable-next-line:no-any
        return !!request.__glimpse_inProcessServerRequest;
    };
    HttpHelper.isInProcessServerResponse = function (response) {
        // tslint:disable-next-line:no-any
        return !!response.__glimpse_inProcessServerResponse;
    };
    HttpHelper.getMaxBodySize = function (configSettings) {
        return configSettings.get('inspector.http.body.capture.maxsize', 132000);
    };
    /**
     * Creates the `body` property for the `data-http-request`, `data-http-response`,
     * `web-request`, and `web-response` messages. The value returned from this
     * method can be directly assigned to the body property of a message
     *
     * @parameter {http.IncomingMessage | http.ClientRequest | http.ServerResponse} request -
     *      The request or response of an HTTP client or server request associated
     *      with the body
     * @parameter {Array<Buffer | string>} bodyChunks - The body chunks, as sent
     *      to `data` events but also truncated so that the length is not greater
     *      than the maximum body capture size configuration option
     * @parameter {number} size - The size of the captured body, irrespective of
     *      the maximum capture body size configuration option
     * @returns {IBodyProperty} The body property for the http message
     */
    HttpHelper.createMessageBodyProperty = function (request, bodyChunks, size, configSettings, partSummaries) {
        // We determine if the body content-type is one that we know how to
        // parse in the client or not, and then change the body property
        // based on this knowledge. Note: the headers exist in different places
        // in the request and response, and sometimes at different times in the
        // request, so we explicitly check headers in a specific order below.
        // tslint:disable-next-line:no-any
        var headers = request.headers || request._headers;
        var contentType = headers['Content-Type'] || headers['content-type'];
        var shouldCaptureBody = false;
        if (contentType) {
            for (var _i = 0, MIME_TYPES_TO_CAPTURE_1 = MIME_TYPES_TO_CAPTURE; _i < MIME_TYPES_TO_CAPTURE_1.length; _i++) {
                var mimeType = MIME_TYPES_TO_CAPTURE_1[_i];
                if (mimeType.test(contentType)) {
                    shouldCaptureBody = true;
                    break;
                }
            }
        }
        // We only create a body property if body content was captured
        if (bodyChunks.length) {
            // Create the body property if we determined that we know how
            // to process the content type
            if (shouldCaptureBody) {
                // TODO: https://github.com/Glimpse/Glimpse.Node/issues/307 Add support for base64 encoding
                var convertedBody = [];
                for (var i = 0; i < bodyChunks.length; i++) {
                    if (Buffer.isBuffer(bodyChunks[i])) {
                        convertedBody.push(bodyChunks[i].toString());
                    }
                    else {
                        convertedBody.push(bodyChunks[i]);
                    }
                }
                var serializedBody = convertedBody.join('');
                if (serializedBody.length > size) {
                    throw new Error('Internal error: size is smaller than body length');
                }
                return {
                    size: size,
                    encoding: 'utf8',
                    content: serializedBody,
                    isTruncated: size > HttpHelper.getMaxBodySize(configSettings),
                    parts: partSummaries
                };
            }
            else {
                return {
                    size: size,
                    encoding: 'none',
                    content: '',
                    isTruncated: true,
                    parts: partSummaries
                };
            }
        }
        // If we got here, this means no body was captured and we want the `body`
        // property in the message to not be included. This return isn't necessary
        // from a JavaScript perspective, since no return is implicitly `return undefined`
        // but we do it explicitly as a reminder of what's going on and why.
        return undefined;
    };
    /**
     * Given an IContentType, return true if the content type is multipart/form-data, false otherwise
     */
    HttpHelper.isMultiPartFormData = function (contentType) {
        return contentType.type === 'multipart' && contentType.subtype === 'form-data';
    };
    /**
     * Given an IContentType, return the boundary delimiter string if available, or undefined if not available.
     * This will strip leading/trailing double-quotes if present.
     */
    HttpHelper.getMultiPartFormBoundary = function (contentType) {
        //tslint:disable-next-line:no-string-literal
        var boundary = contentType.parameters['boundary'];
        if (boundary) {
            if (boundary.indexOf('"') === 0 && boundary.lastIndexOf('"') === boundary.length - 1) {
                // boundary value is enclosed in double-quotes so strip them out
                boundary = boundary.substring(1, boundary.length - 1);
            }
        }
        return boundary;
    };
    HttpHelper.isValidBoundary = function (boundary) {
        // boundary must be between 0 & 70 characters
        return boundary && boundary.length > 0 && boundary.length < 71;
    };
    /**
     * parse a content-type header into its respective parts per RFC 1341
     * https://www.w3.org/Protocols/rfc1341/4_Content-Type.html.
     *
     * Note this implementation is "accomodating" of input that doesn't conform to the spec.
     */
    HttpHelper.parseContentType = function (header) {
        var type = '';
        var subtype = '';
        var parameters = {};
        if (header) {
            // find first slash & set type
            var curr = 0;
            while (curr < header.length && header[curr] !== '/') {
                ++curr;
            }
            var slash = curr;
            // type is case insensitive, so force it to lower case here
            type = header.substring(0, curr).trim().toLowerCase();
            // find first semi-column & set sub-type
            while (curr < header.length && header[curr] !== ';') {
                ++curr;
            }
            var lastSemi = curr;
            if (curr <= header.length) {
                // sub-type is case insensitive
                subtype = header.substring(slash + 1, curr).trim().toLowerCase();
            }
            // parse name/value pairs of parameters section
            ++curr;
            while (curr < header.length) {
                while (curr < header.length && (header[curr] !== '=' && header[curr] !== ';')) {
                    curr++;
                }
                var nameEnd = curr;
                var valueEnd = curr;
                if (header[curr] === '=') {
                    while (curr < header.length && header[curr] !== ';') {
                        curr++;
                    }
                    valueEnd = curr;
                }
                var name_1 = header.substring(lastSemi + 1, nameEnd);
                var value = valueEnd > nameEnd ? header.substring(nameEnd + 1, valueEnd) : '';
                if (name_1.length > 0) {
                    // names are case-insensitive, but values are generally not
                    parameters[name_1.trim().toLowerCase()] = value.trim();
                }
                lastSemi = valueEnd;
                ++curr;
            }
        }
        return {
            type: type,
            subtype: subtype,
            parameters: parameters
        };
    };
    return HttpHelper;
}());
exports.HttpHelper = HttpHelper;
