// @glimpse/glimpse-agent-node/release/tracing/module_instrumentors/MongoDBEvents.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENT_MONGODB_COLLECTION_INSERT = 'invoke|pre|module:/mongodb.collection#insert';
exports.EVENT_MONGODB_COLLECTION_INSERT_METHODS = 'invoke|post|module:/mongodb.collection#insertMethods';
exports.EVENT_MONGODB_COLLECTION_DELETE_METHODS = 'invoke|post|module:/mongodb.collection#deleteMethods';
exports.EVENT_MONGODB_COLLECTION_UPDATE_METHODS = 'invoke|post|module:/mongodb.collection#updateMethods';
exports.EVENT_MONGODB_COLLECTION_COUNT = 'invoke|post|module:/mongodb.collection#count';
exports.EVENT_MONGODB_READ_START = 'notify|Module:/mongodb#startRead';
exports.EVENT_MONGODB_READ_RECORD = 'notify|Module:/mongodb#readRecord';
exports.EVENT_MONGODB_READ_END = 'notify|Module:/mongodb#endRead';
exports.EVENT_MONGODB_CORE_INVOKE_PRE_MONGODB_CORE_SERVER_COMMAND = 'invoke|pre|module:/mongodb-core.server#command';
exports.EVENT_MONGODB_CORE_INVOKE_PRE_MONGODB_CORE_POOL_WRITE = 'invoke|pre|module:/mongodb-core.pool#write';
exports.EVENT_MONGODB_CORE_INVOKE_PRE_MONGODB_CORE_CURSOR_NEXT = 'invoke|pre|module:/mongodb-core.cursor#next';
